google/cloud/forseti/scanner/audit/enabled_apis_rules_engine.py

```diff
@@ -0,0 +1,401 @@
+                for resource_id in resource_ids:
+                    if resource_id == '*':
+                        # Wild-card rules apply to all projects.
+                        gcp_resource = resource_util.create_resource(
+                            resource_id='*', resource_type='project')
+                    else:
+                        gcp_resource = resource_util.create_resource(
+                            resource_id=resource_id,
+                            resource_type=resource_type)
+
+                    rule_def_resource = {
+                        'services': services,
+                        'mode': mode,
+                    }
+                    rule = Rule(rule_name=rule_def.get('name'),
+                                rule_index=rule_index,
+                                rules=rule_def_resource)
+
+                    # If no mapping exists, create it. If the rule isn't in the
+                    # mapping, add it.
+                    self.resource_rules_map[gcp_resource].add(rule)
+
+        finally:
+            self._rules_sema.release()
+
+    def find_violations(self, project, enabled_apis):
+        """Find enabled APIs violations in the rule book.
+
+        Args:
+            project (gcp_type): The project that these APIs are enabled on.
+            enabled_apis (list): list of enabled APIs.
+
+        Returns:
+            iterable: A generator of the rule violations.
+        """
+        violations = itertools.chain()
+
+        # Check for rules on all ancestors, and the wildcard rule.
+        resource_ancestors = (
+            relationship.find_ancestors(project, project.full_name))
+        resource_ancestors.append(resource_util.create_resource(
+            resource_id='*', resource_type='project'))
+
+        for curr_resource in resource_ancestors:
+            resource_rules = self.resource_rules_map.get(curr_resource, [])
+            for resource_rule in resource_rules:
+                violations = itertools.chain(
+                    violations,
+                    resource_rule.find_policy_violations(project, enabled_apis))
+
+        return violations
+
+
+# Mapping from rule mode to mode check method.
+_RULE_MODE_METHODS = {
+    _WHITELIST: _check_whitelist_apis,
+    _BLACKLIST: _check_blacklist_apis,
+    _REQUIRED: _check_required_apis,
+}
+
+
+class Rule(object):
+    """Rule properties from the rule definition file. Also finds violations."""
+
+    RuleViolation = collections.namedtuple(
+        'RuleViolation',
+        ['resource_type', 'resource_id', 'full_name', 'rule_name', 'rule_index',
+         'violation_type', 'apis', 'resource_data'])
+
+    def __init__(self, rule_name, rule_index, rules):
+        """Initialize.
+        Args:
+            rule_name (str): Name of the loaded rule.
+            rule_index (int): The index of the rule from the rule definitions.
+            rules (dict): The rules from the file.
```

Comment:
I think this is supposed to be a rule instead of rules?

Reply:
Fixed!